server/autolinkapp/app.go

```diff
@@ -0,0 +1,172 @@
+
+	editLinksRoute        = "/edit-links"
+	editLinksCommandRoute = "/edit-links-command/submit"
+)
+
+func RegisterHandler(root *mux.Router, store Store, service Service) {
+	a := &app{
+		store:   store,
+		service: service,
+	}
+
+	appRouter := root.PathPrefix(appRoute).Subrouter()
+	appRouter.HandleFunc(manifestRoute, a.handleManifest).Methods(http.MethodGet)
+
+	adminRoutes := appRouter.PathPrefix("").Subrouter()
+	adminRoutes.Use(a.adminRequired)
+
+	adminRoutes.HandleFunc(bindingsRoute, a.handleBindings).Methods(http.MethodPost)
+	adminRoutes.HandleFunc(editLinksRoute+"/form", a.handleFormFetch).Methods(http.MethodPost)
+	adminRoutes.HandleFunc(editLinksRoute+"/submit", a.handleFormSubmit).Methods(http.MethodPost)
+	adminRoutes.HandleFunc(editLinksCommandRoute+"/submit", a.handleCommandSubmit).Methods(http.MethodPost)
+}
+
+func (a *app) handleManifest(w http.ResponseWriter, r *http.Request) {
+	pluginURL := a.service.GetPluginURL()
+	rootURL := pluginURL + appRoute
+
+	manifest := apps.Manifest{
+		AppID:       appID,
+		DisplayName: appDisplayName,
+		AppType:     apps.AppTypeHTTP,
+		HTTPRootURL: rootURL,
+		RequestedPermissions: apps.Permissions{
+			apps.PermissionActAsBot,
+		},
+		RequestedLocations: apps.Locations{
+			apps.LocationCommand,
+		},
+	}
+
+	httputils.WriteJSON(w, manifest)
+}
+
+func (a *app) handleBindings(w http.ResponseWriter, r *http.Request) {
+	call := &apps.CallRequest{}
+	err := json.NewDecoder(r.Body).Decode(call)
+	if err != nil {
+		httputils.WriteJSON(w, apps.NewErrorCallResponse(errors.Wrap(err, "failed to decode call request body")))
+		return
+	}
+
+	icon := a.service.GetPluginURL() + iconRoute
+	resp := &apps.CallResponse{
+		Type: apps.CallResponseTypeOK,
+		Data: []apps.Binding{
+			{
+				Location: apps.LocationCommand,
+				Bindings: []*apps.Binding{
+					{
+						Icon:        icon,
+						Label:       autolinkCommand,
+						Location:    autolinkCommand,
+						Description: appDisplayName,
+						Hint:        "[edit]",
+						Bindings: []*apps.Binding{
+							{
+								Location: "edit",
+								Label:    "edit",
+								Form:     &apps.Form{Fields: []*apps.Field{}},
+								Call: &apps.Call{
+									Path: editLinksCommandRoute,
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	httputils.WriteJSON(w, resp)
+}
+
+func (a *app) handleCommandSubmit(w http.ResponseWriter, r *http.Request) {
+	resp := a.getEmptyFormResponse()
+	httputils.WriteJSON(w, resp)
+}
+
+func (a *app) handleIcon(w http.ResponseWriter, r *http.Request) {
+	resp := a.getEmptyFormResponse()
+	httputils.WriteJSON(w, resp)
+}
+
+func (a *app) adminRequired(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		b, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			httputils.WriteJSON(w, apps.NewErrorCallResponse(errors.Wrap(err, "failed to read call request body")))
+			return
+		}
+
+		r.Body.Close()
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(b))
```

Comment:
Why is this needed?

Reply:
I'm reading from the body in this middleware, and want the http handlers to be able to read from the body as well. I don't like this solution